cli: Improve TypeScript codegen

ValeriaVG opened the pull request:
- Allows Client methods overload
- Provides support for Raw endpoints

**Breaking changes:**
- Generated TypeScript client methods do not throw an error anymore and return `Result` instead, which either contains `{error: ErrorResponse}` or `data` field with the json from the response body
- Raw endpoint methods return underlying `Response`, without consuming response body
- Void endpoints no longer consume response body, unless they result in an error

Tested on JS & Deno; in order to support NodeJS one needs to either import fetch polyfill or extend `Client` class and override `doRaw` method.

Closes #114,#50

ValeriaVG commented:
I've generated a client for a pet project [ValeriaVG/social-network](https://github.com/ValeriaVG/social-network/blob/main/pwa/src/lib/api/autogen.ts). The resulting code is:

```ts
export interface ErrorResponse {
    code: number,
    message: string,
    details: any
}

export type Result<T> = { data: T } | { error: ErrorResponse }

export default class Client {
    follow: follow.ServiceClient
    post: post.ServiceClient
    user: user.ServiceClient

    public async doRaw(method: string, path: string, body?: any): Promise<Response> {
        const headers: Record<string,string> = { "Content-Type": "application/json" }
        if (this.token) {
            headers["Authorization"] = "Bearer " +this.token
        }
        return fetch(this.baseURL + path, {
            method,
            headers,
            body
        })
    }

    public async do<T>(method: string, path: string, req?: any): Promise<Result<T>> {
        try {
            const response = await this.doRaw(method, path, req !== undefined ? JSON.stringify(req) : undefined)
            if (!response.ok) {
                const error = <ErrorResponse>(await response.json())
                return { error }
            }
            return { data: <T>(await response.json().catch(_ => null)) }
        } catch (error) {
            return {
                error: <ErrorResponse>{
                    code: -1,
                    message: error.message
                }
            }
        }
    }

    public async doVoid(method: string, path: string, req?: any): Promise<ErrorResponse | null> {
        try {
            const response = await this.doRaw(method, path, req !== undefined ? JSON.stringify(req) : undefined)
            if (!response.ok) {
                const error = <ErrorResponse>(await response.json())
                return error
            }
            return null

        } catch (error) {
            return <ErrorResponse>{
                code: -1,
                message: error.message

            }
        }
    }

    protected baseURL: string

    constructor(environment: string = "prod", public token?: string) {
        this.baseURL = environment === "local"? "http://localhost:4000": `https://social-network-qsbi.encoreapi.com/${environment}`
        this.follow = new follow.ServiceClient(this)
        this.post = new post.ServiceClient(this)
        this.user = new user.ServiceClient(this)
    }
}

export namespace follow {
    export interface FollowingState {
        Following: boolean
    }

    export class ServiceClient {
        private client: Client

        constructor(client: Client) {
            this.client = client
        }

        public FollowTag(tag: string) {
            return this.client.doVoid("PUT", `/follow/tag/${tag}`)
        }

        public FollowUser(username: string) {
            return this.client.doVoid("PUT", `/follow/username/${username}`)
        }

        public IsFollowingTag(tag: string) {
            return this.client.do<FollowingState>("GET", `/following/tag/${tag}`)
        }

        public IsFollowingUser(username: string) {
            return this.client.do<FollowingState>("GET", `/following/username/${username}`)
        }

        public UnfollowTag(tag: string) {
            return this.client.doVoid("PUT", `/unfollow/tag/${tag}`)
        }

        public UnfollowUser(username: string) {
            return this.client.doVoid("PUT", `/unfollow/username/${username}`)
        }
    }
}

export namespace post {
    export interface Author {
        Username: string
        Avatar: string
    }

    export type MediaType = string

    export interface Post {
        ID: string
        Author: Author
        Content: string
        Media: PostMedia[]
        CreatedAt: string
    }

    export interface PostForm {
        Content: string
        Media: PostMedia[]
    }

    export interface PostList {
        Items: Post[]
        Total: number
        Next: number
    }

    export interface PostMedia {
        /**
         * Type can be "audio","image" or "video"
         */
        Type: MediaType

        /**
         * URL is a link to a media source
         */
        URL: string

        /**
         * Width is width of image or video frame
         */
        Width: number

        /**
         * Height is height of image or video frame
         */
        Height: number

        /**
         * Preview is an image that used as cover
         */
        Preview: string
    }

    export class ServiceClient {
        private client: Client

        constructor(client: Client) {
            this.client = client
        }

        /**
         * CreatePost creates a media post with current user as an author
         */
        public CreatePost(params: PostForm) {
            return this.client.do<Post>("POST", `/post`, params)
        }

        /**
         * GetPost retrieves post with provided id
         */
        public GetPost(id: string) {
            return this.client.do<Post>("GET", `/post/${id}`)
        }

        /**
         * ListTaggedPosts lists posts with a provided tag
         */
        public ListTaggedPosts(tag: string, params: common.ListParams) {
            const query: any[] = [
                "limit", params.Limit,
                "created_before", params.CreatedBefore,
            ]
            return this.client.do<PostList>("GET", `/posts/tag/${tag}?${encodeQuery(query)}`)
        }

        /**
         * ListUserFeed lists posts, relevant for a current user
         */
        public ListUserFeed(params: common.ListParams) {
            const query: any[] = [
                "limit", params.Limit,
                "created_before", params.CreatedBefore,
            ]
            return this.client.do<PostList>("GET", `/feed?${encodeQuery(query)}`)
        }

        /**
         * ListUserPosts lists user posts with a provided username
         */
        public ListUserPosts(username: string, params: common.ListParams) {
            const query: any[] = [
                "limit", params.Limit,
                "created_before", params.CreatedBefore,
            ]
            return this.client.do<PostList>("GET", `/posts/author/${username}?${encodeQuery(query)}`)
        }
    }
}

export namespace user {
    export interface AuthorizedUserResponse {
        User: User
        Token: string
    }

    export interface Profile {
        Username: string
        Avatar: string
        Info: string
    }

    export interface ProfileForm {
        Avatar: string
        Info: string
    }

    export interface RegisterForm {
        Email: string
        Username: string
        Password: string
    }

    export interface User {
        ID: string
        Email: string
        Username: string
        CreatedAt: string
    }

    export interface UserCredentials {
        Email: string
        Password: string
    }

    export class ServiceClient {
        private client: Client

        constructor(client: Client) {
            this.client = client
        }

        /**
         * GetUserProfile returns profile information of user with provided username
         */
        public GetUserProfile(username: string) {
            return this.client.do<Profile>("GET", `/user/profile/username/${username}`)
        }

        /**
         * Login authenticates user with the provided credentials
         */
        public Login(params: UserCredentials) {
            return this.client.do<AuthorizedUserResponse>("POST", `/user/login`, params)
        }

        /**
         * Me returns current authenticated user
         */
        public Me() {
            return this.client.do<AuthorizedUserResponse>("GET", `/user/me`)
        }

        /**
         * Register creates new user with the provided credentials
         */
        public Register(params: RegisterForm) {
            return this.client.do<AuthorizedUserResponse>("POST", `/user/register`, params)
        }

        /**
         * UpdateProfile updates profile information of a current user
         */
        public UpdateProfile(params: ProfileForm) {
            return this.client.do<Profile>("PUT", `/user/profile`, params)
        }
    }
}

export namespace common {
    export interface ListParams {
        Limit: number
        CreatedBefore: number
    }
}

function encodeQuery(parts: any[]): string {
    const pairs = []
    for (let i = 0; i < parts.length; i += 2) {
        const key = parts[i]
        let val = parts[i+1]
        if (!Array.isArray(val)) {
            val = [val]
        }
        for (const v of val) {
            pairs.push(`${key}=${encodeURIComponent(v)}`)
        }
    }
    return pairs.join("&")
}
```

alexisvisco commented on cli/internal/codegen/ts.go:
```diff
@@ -307,6 +307,15 @@ func (ts *ts) writeDeclDef(ns string, decl *schema.Decl) {
+	ts.WriteString("    message: string,\n")
+	ts.WriteString("    details: any\n")
+	ts.WriteString("}\n\n")
+	ts.WriteString("export type Result<T> = { data: T } | { error: ErrorResponse }\n\n")
```
Isn't a problem to have to do properties assertion to infer result type ?

ValeriaVG replied:
@alexisvisco It shouldn't be, in TypeScript it is quite easy to it with:
```ts
if('error' in response){
 // handle error case
}
```
Besides I believe that it's a good way to not forget to handle the errors. What do you think?

ValeriaVG commented:
- Rebased on main branch
- Added ability to provide URL instead of environment:
```ts
if (environment.startsWith('http://') || environment.startsWith('https://')) {
      this.baseURL = environment
} else {
      this.baseURL = environment === "local" ? "http://localhost:4000" : `https://app.encoreapi.com/${environment}`
}
```

DomBlack commented:
@ValeriaVG I'm going to close this now, as I believe all the changes you proposed in this PR have been merged in the updated client generation for Encore 1.1.0.

Feel free to re-open if you think something is missing from this PR 😄
